cmd/bad-key-revoker/main.go

```diff
@@ -463,19 +463,23 @@ func main() {
 		emailTemplate:   emailTemplate,
 		logger:          logger,
 	}
+
+	// Don't loop faster than our interval
+	min_iter_speed_ticker := time.NewTicker(config.BadKeyRevoker.Interval.Duration)
+
 	for {
 		noWork, err := bkr.invoke()
 		if err != nil {
 			keysProcessed.WithLabelValues("error").Inc()
 			logger.AuditErrf("failed to process blockedKeys row: %s", err)
-			continue
-		}
-		if noWork {
-			logger.Info(fmt.Sprintf(
-				"No work to do. Sleeping for %s", config.BadKeyRevoker.Interval.Duration))
-			time.Sleep(config.BadKeyRevoker.Interval.Duration)
 		} else {
-			keysProcessed.WithLabelValues("success").Inc()
+			if noWork {
+				logger.Info(fmt.Sprintf(
+					"No work to do. Sleeping for %s", config.BadKeyRevoker.Interval.Duration))
```

Comment:
This log line is no longer accurate, for two reasons:
1) it's not sleeping for X duration, it's sleeping until the next X duration tick
2) we now _also_ sleep even if there is more work to do

It seems like what you're trying to do here is to make it so that errors don't cause us to loop immediately, but you've also made it so that success also doesn't cause us to loop immediately. I can imagine a world in which this causes bkr to consistently fall behind (e.g. in case of a very large key compromise disclosure).

I think that making this only sleep if it has no work left or if it errors out is definitely harder; it's a larger refactor for sure. But it may be worth doing so this doesn't inadvertently change "normal" behavior too.

Reply:
Indeed. That said, do we actually want it to tight-loop? From an operations side, I'd rather set the minimum interval to 1s and know it won't run a large select query more often than that, which then is tunable.

It's worth pointing out that this routine isn't slowing down the inner loop for `findUnrevoked`, which handles all the actual batching, if there's a large number of keys to revoke at once. So this doesn't slow down a large batch. Rather, it effectively makes the utility behave as systemd's `OnUnitActiveSec` timer in predictability.

Happy to edit the misleading log message, though, and I'm OK with abandoning this change, but it is deliberate as-is.

Reply:
Ugh, I take that back, this is not an intuitive tool.